jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java

```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
           FileSystems.getDefault().getPathMatcher("glob:" + entry.getKey()), entry.getValue());
     }
 
-    new DirectoryWalker(sourceDirectory)
-        .filterRoot()
-        .walk(
-            localPath -> {
-              AbsoluteUnixPath pathOnContainer =
-                  targetDirectory.resolve(sourceDirectory.relativize(localPath));
-              Instant modificationTime = modificationTimeProvider.get(localPath, pathOnContainer);
-              FilePermissions permissions =
-                  extraDirectoryPermissions.get(pathOnContainer.toString());
-              if (permissions == null) {
-                // Check for matching globs
-                Path containerPath = Paths.get(pathOnContainer.toString());
-                for (Map.Entry<PathMatcher, FilePermissions> entry : pathMatchers.entrySet()) {
-                  if (entry.getKey().matches(containerPath)) {
-                    builder.addEntry(
-                        localPath, pathOnContainer, entry.getValue(), modificationTime);
-                    return;
-                  }
-                }
-
-                // Add with default permissions
-                builder.addEntry(localPath, pathOnContainer, modificationTime);
-              } else {
-                // Add with explicit permissions
-                builder.addEntry(localPath, pathOnContainer, permissions, modificationTime);
+    DirectoryWalker walker = new DirectoryWalker(sourceDirectory).filterRoot();
+    excludes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .forEach(pathMatcher -> walker.filter(path -> !pathMatcher.matches(path)));
+    includes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .map(pathMatcher -> (Predicate<Path>) (path -> pathMatcher.matches(path)))
+        .reduce((matches1, matches2) -> matches1.or(matches2))
+        .ifPresent(walker::filter);
```

Comment:
I'm probably missing something here, why is the `reduce` step needed here?

Reply:
So before `reduce`, we have a list of `Predicate`s where each predicate tests if a given path matches its `<include>` glob pattern (i.e., testing if the file should be included).

Then `reduce` converts a list to a single `Predicate` that is contributed by "or"ing all the predicates. So, this predicate tests if a given path matches any of the `<include>` patterns. And `walker::filter()` accepts a single predicate. Note calling `walker::filter` repeatedly for each `Predicate` before `reduce()` makes them connected with "and", not "or".

Reply:
Makes sense. Thanks for the explanation!